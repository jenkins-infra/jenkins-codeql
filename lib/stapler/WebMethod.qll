import java

import jenkins.ProductionCallable

class WebMethod extends Method, ProductionCallable {
    WebMethod() {
        // TODO make this nicer and more comprehensive, e.g. all parameters need to have one of several annotations or be one of several types

        // Stapler basic requirements for a method to be dispatchable
        // and include Jenkins DoActionFilter since 2.138.3
        this.isPublic()
        and
        not(this.hasAnnotation("jenkins.security.stapler", "StaplerNotDispatchable"))
        and
        (
            // Ensure the web method is properly annotated
            this.hasAnnotation("jenkins.security.stapler", "StaplerDispatchable")
            or
            this.hasAnnotation("org.kohsuke.stapler", "WebMethod")
            or
            this.getAnAnnotation().getType().hasAnnotation("org.kohsuke.stapler.interceptor", "InterceptorAnnotation")
            or
            this.getName().indexOf("do") = 0
        )
        and not exists(Parameter p | this.getAParameter() = p
            // Ensure no parameter exists without expected type or annotation
            and
            (
                not p.getType().hasName("StaplerRequest")
                and not p.getType().hasName("StaplerResponse")
                and not p.getType().hasName("HttpServletRequest")
                and not p.getType().hasName("HttpServletResponse")
                and not p.getType().hasName("ServletRequest")
                and not p.getType().hasName("ServletResponse")
                and not p.getType().hasName("RequestImpl")
                and not p.getType().hasName("ResponseImpl")
            )
            and not exists(Annotation a | p.getAnAnnotation() = a
                and
                (
                    a.getType().hasName("QueryParameter")
                    or
                    a.getType().hasName("AncestorInPath")
                    or
                    a.getType().hasName("Header")
                    or
                    a.getType().hasName("JsonBody")
                    or
                    a.getType().hasName("SubmittedForm")
                )
            )
        )
    }
}
