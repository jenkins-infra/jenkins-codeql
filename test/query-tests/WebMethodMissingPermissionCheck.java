import org.kohsuke.stapler.verb.POST;
import org.kohsuke.stapler.interceptor.RequirePOST;
import org.kohsuke.stapler.QueryParameter;
import org.kohsuke.stapler.StaplerRequest;

import java.net.URL;

public class WebMethodMissingPermissionCheck {
    @POST
    public void doWhatever(@QueryParameter String foo) throws Exception {
        new URL("value").openConnection();
    }

    @RequirePOST
    public void doWhatever2(StaplerRequest req) throws Exception {
        new URL("value").openConnection();
    }
}