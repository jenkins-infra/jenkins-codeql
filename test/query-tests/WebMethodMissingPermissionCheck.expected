| FileFromRestOfPath.java:7:17:7:31 | doPathTraversal | Potential missing permission check in FileFromRestOfPath#doPathTraversal |
| FileFromRestOfPath.java:12:19:12:27 | doNotCare | Potential missing permission check in FileFromRestOfPath#doNotCare |
| WebMethodMissingPermissionCheck.java:10:17:10:26 | doWhatever | Potential missing permission check in WebMethodMissingPermissionCheck#doWhatever |
| WebMethodMissingPermissionCheck.java:15:17:15:27 | doWhatever2 | Potential missing permission check in WebMethodMissingPermissionCheck#doWhatever2 |
